File: 10-逆向工程/03-协议逆向/协议状态机恢复/pfsm.go
// 协议状态机恢复：Veritas(ACNS 2011) P-PSM 骨架实现。
// 流水线: 消息类型标注(首token+方向) → 逐流转移计数(不跨流) → 概率转移+阈值剪枝 → PFSM。
// 用法: go run pfsm.go   (自检失败即 panic)
package main

import (
	"fmt"
	"math/rand"
	"sort"
	"strings"
)

// ---------- 1. 消息类型标注 ----------
type msg struct {
	dir  byte // 'C' / 'S'
	text string
}

type state struct {
	dir  byte
	typ  string
}

func (s state) String() string { return string(s.dir) + ":" + s.typ }

func label(m msg) state {
	fields := strings.Fields(m.text)
	t := ""
	if len(fields) > 0 {
		t = strings.ToUpper(fields[0])
	}
	return state{m.dir, t}
}

func labelFlow(flow []msg) []state {
	out := make([]state, len(flow))
	for i, m := range flow {
		out[i] = label(m)
	}
	return out
}

// ---------- 2. 转移计数 (逐流, 禁止跨流) ----------
type edgeKey struct{ s, d state }

func transitions(flows [][]msg) map[edgeKey]int {
	C := map[edgeKey]int{}
	for _, flow := range flows {
		seq := labelFlow(flow)
		for i := 0; i+1 < len(seq); i++ {
			C[edgeKey{seq[i], seq[i+1]}]++
		}
	}
	return C
}

// ---------- 3. 概率化 + 剪枝 ----------
type edge struct {
	s, d  state
	count int
	prob  float64
}

func pfsm(C map[edgeKey]int, pMin float64, cMin int) []edge {
	rowTot := map[state]int{}
	for k, c := range C {
		rowTot[k.s] += c
	}
	var edges []edge
	for k, c := range C {
		p := float64(c) / float64(rowTot[k.s])
		if c >= cMin && p >= pMin {
			edges = append(edges, edge{k.s, k.d, c, p})
		}
	}
	sort.Slice(edges, func(i, j int) bool {
		if edges[i].prob != edges[j].prob {
			return edges[i].prob > edges[j].prob
		}
		if edges[i].s.String() != edges[j].s.String() {
			return edges[i].s.String() < edges[j].s.String()
		}
		return edges[i].d.String() < edges[j].d.String()
	})
	return edges
}

func accepts(edges []edge, seq []state) bool {
	ok := map[edgeKey]bool{}
	for _, e := range edges {
		ok[edgeKey{e.s, e.d}] = true
	}
	for i := 0; i+1 < len(seq); i++ {
		if !ok[edgeKey{seq[i], seq[i+1]}] {
			return false
		}
	}
	return true
}

// ---------- 自检 ----------
func selfTest() {
	flows := [][]msg{
		{{'C', "A"}, {'S', "B"}, {'C', "A"}},
		{{'C', "A"}, {'S', "B"}, {'C', "C"}},
	}
	C := transitions(flows)
	if C[edgeKey{{'C', "A"}, {'S', "B"}}] != 2 {
		panic("count wrong")
	}
	// 跨流伪边: 流1结尾 ('C','A') 与 流2开头 ('C','A') 不得产生转移
	if _, ok := C[edgeKey{{'C', "A"}, {'C', "C"}}]; ok {
		panic("cross-flow transition leaked")
	}
	edges := pfsm(C, 0.05, 1)
	found := false
	for _, e := range edges {
		if e.s.String() == "C:A" && e.d.String() == "S:B" && e.count == 2 && e.prob == 1.0 {
			found = true
		}
	}
	if !found {
		panic("expected edge missing")
	}
	// 剪枝: 低频低概率边被删
	C2 := transitions([][]msg{
		{{'X', "A"}, {'X', "B"}},
	})
	for i := 0; i < 100; i++ {
		C2[edgeKey{{'X', "A"}, {'X', "A"}}]++
	}
	for _, e := range pfsm(C2, 0.05, 1) {
		if e.s.typ == "A" && e.d.typ == "B" {
			panic("low-prob edge not pruned")
		}
	}
}

// ---------- 演示: 合成 SMTP 风格流量 ----------
func genSMTPFlows(nFlows, seed int) [][]msg {
	rng := rand.New(rand.NewSource(int64(seed)))
	var flows [][]msg
	for i := 0; i < nFlows; i++ {
		var f []msg
		add := func(d byte, s string) { f = append(f, msg{d, s}) }
		add('C', "EHLO client.example"); add('S', "250-STARTTLS")
		add('C', "MAIL FROM:<a@x>"); add('S', "250 OK")
		for j := 0; j < 1+rng.Intn(3); j++ {
			add('C', "RCPT TO:<b@y>"); add('S', "250 OK")
			if rng.Float64() < 0.2 {
				add('C', "RCPT TO:<bad@z>"); add('S', "550 No such user")
			}
		}
		add('C', "DATA"); add('S', "354 Go ahead")
		add('C', "body line 1"); add('C', "."); add('S', "250 Accepted")
		if rng.Float64() < 0.15 {
			add('C', "EHLO again"); add('S', "250-STARTTLS")
		}
		add('C', "QUIT"); add('S', "221 Bye")
		flows = append(flows, f)
	}
	return flows
}

func main() {
	selfTest()
	flows := genSMTPFlows(30, 1)
	fmt.Println("== 样本: 第一条流的消息类型序列 ==")
	for i, s := range labelFlow(flows[0]) {
		if i > 0 {
			fmt.Print(" -> ")
		}
		fmt.Print(s)
	}
	fmt.Println()
	fmt.Println()
	edges := pfsm(transitions(flows), 0.05, 3)
	fmt.Printf("== 剪枝后 PFSM 边表 (p>=0.05 且 count>=3, 共 %d 条边) ==\n", len(edges))
	for _, e := range edges {
		fmt.Printf("  %-16s -> %-16s  count=%-3d p=%.2f\n", e.s, e.d, e.count, e.prob)
	}
	fmt.Println()
	nOK := 0
	for _, f := range flows {
		if accepts(edges, labelFlow(f)) {
			nOK++
		}
	}
	fmt.Printf("== 流覆盖率回测: %d/%d = %.0f%% ==\n", nOK, len(flows), float64(nOK)/float64(len(flows))*100)
}

File: 10-逆向工程/03-协议逆向/协议状态机恢复/pfsm.py
#!/usr/bin/env python3
"""协议状态机恢复：Veritas(ACNS 2011) P-PSM 骨架实现。

流水线: 消息类型标注(首token+方向) → 逐流转移计数(不跨流) → 概率转移+阈值剪枝 → PFSM。
用法: python pfsm.py   (内置 assert 自检)
"""
from collections import defaultdict

# ---------- 1. 消息类型标注 ----------
def label(msg):
    """(方向, 文本) → 状态标签。方向必须进标签(README 坑1)。"""
    direction, text = msg
    first = text.split(' ', 1)[0].upper()
    return (direction, first)

def label_flow(flow):
    return [label(m) for m in flow]

# ---------- 2. 转移计数 (逐流统计, 禁止跨流边界) ----------
def transitions(flows):
    """flows = [[(dir, text), ...], ...] → 嵌套计数 {src: {dst: count}}。"""
    C = defaultdict(lambda: defaultdict(int))
    for flow in flows:
        seq = label_flow(flow)
        for a, b in zip(seq, seq[1:]):
            C[a][b] += 1
    return C

# ---------- 3. 概率化 + 剪枝 ----------
def pfsm(C, p_min=0.05, c_min=3):
    """转移计数 → 剪枝后的边表 [(src, dst, count, prob)]。"""
    edges = []
    for s, row in C.items():
        tot = sum(row.values())
        for s2, c in row.items():
            if c >= c_min and c / tot >= p_min:
                edges.append((s, s2, c, c / tot))
    return sorted(edges, key=lambda e: (-e[3], e[0], e[1]))

def accepts(edges, seq):
    """流覆盖率回测: 每条相邻转移是否都在剪枝后的 PFSM 边集里。"""
    ok = {(s, d) for s, d, _, _ in edges}
    return all((a, b) in ok for a, b in zip(seq, seq[1:]))

# ---------- 4. (简化)状态合并 ----------
def merge_states(edges, merge_map):
    return [(merge_map.get(s, s), merge_map.get(d, d), c, p)
            for s, d, c, p in edges]

# ---------- 自检 ----------
def self_test():
    flows = [
        [('C', 'A'), ('S', 'B'), ('C', 'A')],
        [('C', 'A'), ('S', 'B'), ('C', 'C')],
    ]
    C = transitions(flows)
    assert C[('C', 'A')][('S', 'B')] == 2
    assert C[('S', 'B')][('C', 'A')] == 1
    # 跨流伪边: 流1结尾 ('C','A') 与 流2开头 ('C','A') 不得产生转移
    assert ('C', 'C') not in C[('C', 'A')]
    edges = pfsm(C, p_min=0.05, c_min=1)
    assert (('C', 'A'), ('S', 'B'), 2, 1.0) in edges
    # 剪枝: A→B 1 次, p=1/100=0.01 < 0.05 → 剪掉
    C2 = transitions([[('X', 'A'), ('X', 'B')]] + [[('X', 'A')] * 100])
    assert not any(s == ('X', 'A') and d == ('X', 'B') for s, d, _, _ in pfsm(C2))
    # 回测
    assert accepts(edges, [('C', 'A'), ('S', 'B'), ('C', 'A')])
    assert not accepts(edges, [('S', 'B'), ('C', 'Z')])

# ---------- 演示: 合成 SMTP 风格流量 ----------
def gen_smtp_flows(n_flows=30, seed=1):
    import random
    rng = random.Random(seed)
    flows = []
    for _ in range(n_flows):
        f = [('C', 'EHLO client.example'), ('S', '250-STARTTLS')]
        f.append(('C', 'MAIL FROM:<a@x>')); f.append(('S', '250 OK'))
        for _ in range(rng.randrange(1, 4)):            # 1~3 个收件人
            f.append(('C', 'RCPT TO:<b@y>'))
            f.append(('S', '250 OK'))
            if rng.random() < 0.2:                      # 偶发 550 拒收
                f.append(('C', 'RCPT TO:<bad@z>'))
                f.append(('S', '550 No such user'))
        f.append(('C', 'DATA')); f.append(('S', '354 Go ahead'))
        f.append(('C', 'body line 1')); f.append(('C', '.')); f.append(('S', '250 Accepted'))
        if rng.random() < 0.15:                         # 偶发重新 EHLO
            f.append(('C', 'EHLO again')); f.append(('S', '250-STARTTLS'))
        f.append(('C', 'QUIT')); f.append(('S', '221 Bye'))
        flows.append(f)
    return flows

def demo():
    flows = gen_smtp_flows()
    print('== 样本: 第一条流的消息类型序列 ==')
    print(' ', ' -> '.join('%s:%s' % (d, t) for d, t in label_flow(flows[0])))
    print()
    C = transitions(flows)
    edges = pfsm(C)
    print('== 剪枝后 PFSM 边表 (p>=0.05 且 count>=3, 共 %d 条边) ==' % len(edges))
    for s, d, c, p in edges:
        print('  %-16s -> %-16s  count=%-3d p=%.2f' % ('%s:%s' % s, '%s:%s' % d, c, p))
    print()
    # 流覆盖率回测 (Veritas 用类似指标衡量保真度)
    n_ok = sum(accepts(edges, label_flow(f)) for f in flows)
    print('== 流覆盖率回测: %d/%d = %.0f%% ==' % (n_ok, len(flows), n_ok / len(flows) * 100))
    print()
    # 状态合并示例: C:BODY 与 C:. 同属 DATA 阶段客户端内容 → 合并
    mmap = {('C', 'BODY'): ('C', 'DATA_BODY'), ('C', '.'): ('C', 'DATA_BODY')}
    region = [e for e in edges if ('C', 'BODY') in e[:2] or ('C', '.') in e[:2]]
    print('== 状态合并示例: DATA 阶段局部边表 ==')
    print('  合并前:')
    for s, d, c, p in region:
        print('    %-16s -> %-16s  p=%.2f' % ('%s:%s' % s, '%s:%s' % d, p))
    print('  合并后 (C:BODY, C:. → C:DATA_BODY):')
    seen = set()
    for s, d, c, p in merge_states(region, mmap):
        if (s, d) in seen:
            continue
        seen.add((s, d))
        print('    %-16s -> %-16s  p=%.2f' % ('%s:%s' % s, '%s:%s' % d, p))

if __name__ == '__main__':
    self_test()
    demo()
